### ISSUE TITLE
Proposal: re-use http://discuss.ipfs.io as comments engine

### ISSUE THREAD

**Issue by lidel** [ipfs/blog#100]
Here's an idea:

> Discourse has the ability to embed the comments from a topic in a remote site using a Javascript API that creates an IFRAME. For an example of this in action, check out [the bottom of this] [Coding Horror's blog post](http://blog.codinghorror.com/welcome-to-the-internet-of-compromised-things/#discourse-comments). The blog is run via Ghost but the comments are embedded from his Discourse forum.
> 
> One important thing to note with this setup is that users have to navigate to your forum to post replies. This is intentional, as we feel that the posting interface on a Discourse forum is currently much richer than what we could embed via Javascript.
> 
> This guide will show how to set up comment embedding on your own blog or web site.
> 
> https://meta.discourse.org/t/embedding-discourse-comments-via-javascript/31963

**TL;DR** on how it works:

> If you set this up correctly, the first time you visit [http://ipfs.io/blog/entry-123.html] it will try to load comments for the blog post. Since there are none, it will tell the Discourse forum to create a new topic in the background. A new topic will be created by [some account] and the contents of the first post will be crawled from your blog and the text will be extracted automatically.

> Once the new topic is created, users can post on it, and their comments will automatically be displayed the next time [http://ipfs.io/blog/entry-123.html] is visited.

I imagine comment-topics would be created in dedicated category, eg subcategory "Blog" eg. in "Announcements".

I think re-using https://discuss.ipfs.io for blog comments may provide value on multiple levels, for example:

- it is easy to switch to different blog engine without worrying about lost comments or broken formatting
- discussions are not lost in closed "blog engine silo", they will be stored on community forum 
- It provides free "onboarding": adding a comment to a blog post is an easy and safe way to get familiar with Discourse (engine used by https://discuss.ipfs.io)

PS. This is #💯 issue! 🥇

**Comment by lgierth** [comment]
Great idea :+1:

**Comment by VictorBjelkholm** [comment]
One quick thought on this. There are two different ways we could do this. 

1. Generate topics automatically. This can become a hassle if we don't figure out which URL scheme to use for this, as the blog could be visited in many different ways (blog.ipfs.io, ipfs.io/blog, ipfs.io/ipfs/$hashOfBlog, localhost/ipfs/$hashOfBlog) and Discourse could be confused into believing they are all different topics. We could just use the final part of the URL to generate the topic though, although I think the embedding snippet asks for a full URL. 

2. Add a `postID` point to every blogpost where we manually create topics on discuss.ipfs.io and then link them to the blogpost. This makes it a bit harder to maintain but it'll always link to the same topic no matter how you access it.

We also have to figure out a fallback for when you're offline, so we show some warning if that's the case.

**Comment by VictorBjelkholm** [comment]
Went with solution #1, which works well as we set the canonical URL under `blog.ipfs.io` and no matter how you visit it, that'll be the URL that gets linked with the post.

However, embedding only works for whitelisted domains, so embedding won't even try to embed the comments if visiting the post under another domain than localhost:1313 (for dev) or blog.ipfs.io.

PR for implementing this is here: https://github.com/ipfs/blog/pull/159 Reviews are welcome so we can move forward with it :)

**Comment by lidel** [comment]
Somehow related: I noticed that https://ipfs.io/blog/ does not redirect to https://blog.ipfs.io, but hosts content directly.   I imagine https://ipfs.io/blog/ will not have correct comments unless whitelisted?
Shouldn't it redirect to https://blog.ipfs.io, to remove dupplicated URLs?

**Comment by VictorBjelkholm** [comment]
@lidel right, good point. According to https://github.com/ipfs/blog/issues/116 the canonical URL is support to be ipfs.io/blog, not blog.ipfs.io which is supported as legacy right now. So rather blog.ipfs.io should redirect to ipfs.io/blog, which for now just loads blog.ipfs.io but in the future would actually shows the content of /blog from github.com/ipfs/website. Changing the PR to reflect this.

Both ipfs.io/blog and blog.ipfs.io are whitelisted in discuss.ipfs.io to be able to comments.

**Comment by lidel** [comment]
I am looking into making dnslink enabled by default in [ipfs-companion](https://github.com/ipfs/ipfs-companion) and our own websites are not ideal when loaded from IPNS.

I created https://github.com/ipfs/website/pull/270 to default to `blog.ipfs.io` as a quick fix for navigation links when loaded via `/ipns/ipfs.io/`

Question: are we able to whitelist comment embedding into non-http URLs? 
Namely `ipns://blog.ipfs.io`

**Comment by lidel** [comment]
@VictorBjelkholm is it possible to  add `http://127.0.0.1:8080` as a valid embed host at our Discourse instance? It would fix comments for a lot of people using DNSLink :)

**Comment by Stebalien** [comment]
I assume we'll end up with an "insecure content" problem, unfortunately (unless the browser treats localhost differently?).

**Comment by lidel** [comment]
@Stebalien it is safe: browser vendors whitelist `127.0.0.1` and `::1`  and those do not trigger mixed-content warnings.
 
- Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=903966
- Chromium: https://chromium.googlesource.com/chromium/src.git/+/130ee686fa00b617bfc001ceb3bb49782da2cb4e

The formal background on this decision is probably Step 4 from [_Is origin potentially trustworthy?_ in  HTML Spec](https://w3c.github.io/webappsec-secure-contexts/#is-origin-trustworthy).

**Comment by Stebalien** [comment]
Ah, got it.

**Comment by lidel** [comment]
Its done! 

http://blog.ipfs.io/weekly-80/ has a comments section at the bottom, when loaded from original server comments are displayed from topic in https://discuss.ipfs.io/c/news/blog-entries/20

Loading comments when on local IPFS gateway is tracked in https://github.com/ipfs/blog/issues/175
